sbt/src/sbt-test/nio/clean/build.sbt

```diff
@@ -0,0 +1,37 @@
+import sjsonnew.BasicJsonProtocol._
+
+val copyFile = taskKey[Int]("dummy task")
+copyFile / fileInputs += baseDirectory.value.toGlob / "base" / "*.txt"
+copyFile / fileOutputs += baseDirectory.value.toGlob / "out" / "*.txt"
```

Comment:
Does registering `fileOutputs` automatically invalidate the cache?

Reply:
I'm not sure what you mean by the cache. The fileOutputs setting is used in two tasks: the custom clean implementation and the allOutputPaths task defined in sbt.nio.Keys. The allOutputPaths task writes to the previous cache so that we can implement the modifiedOutputPaths, changedOutputPaths, removedOutputPaths and outputFileStamps tasks.

Also both fileInputs and fileOutputs are settings so there shouldn't be side effects in running them.

Reply:
I was thinking of a situation where you have task `foo` whose `foo / fileInputs` point to `baseDirectory.value.toGlob / "out" / "*.txt"`. When `copyFile` runs, would that automatically invalidate the internal stamps, or would `foo` need to poll, or explicitly depend on `copyFile`?

Reply:
This test should be rewritten so that copyFile returns Seq[Path]. Presently, a dependent task would implicitly poll `baseDirectory.value.toGlob / "out" / "*.txt"` if it called `(copyFile / allOutputPaths).value`. If instead, copyFile returned `Seq[Path]` then `fileOutputs` would not need to be specified at all and `(copyFile / allOutputPaths).value` would return the paths from the previous cache for `copyFile.value`.

Reply:
I hadn't yet thought of automatically caching the values of a task returning a collection of files or paths when I wrote this test.

The situation where fileOutputs is useful is when your task delegates to an external task that has the side effect of writing files into an output path but doesn't return the output files. For example if you shell out to invoke the typescript compiler on a set of files, you don't necessarily know what all the output files are but you may know what directory they will be in and what extension they have. So you might do
```
typeScript / fileOutputs += target.value.toGlob / "js" / ** / "*.js"
```
and then you might have another task the minifies that looks like and delegates to some function `def minify(minifiedJS: Path, inputs: Seq[Path])`
```
minififyJS := {
  val output = target.value.toPath / "minified.js"
  minifyJS.previous match {
    case Some(o: Path) if (typeScript / changedOutputPaths).value.isEmpty => o
    case _ => minify(output, (typeScript / allOutputPaths).value)
  }
 }
```
the `(typeScript / allOutputPaths).value` will proxy through (typeScript / fileOutputs). This will poll the file system. The `(typeScript / changedOutputPaths).value` will delegate to `(typeScript / outputFileStamps).value and `(typeScript / outputFileStamps).previous`. The implementation of outputFileStamps delegates to allOutputPaths, so we only ever poll `baseDirectory.value.toGlob / "out" / "*.txt"` once.

Reply:
If `typeScript := taskKey[Seq[Path]]("Compiles input typescript files and returns the generated js files")`, then there would be no reason to define `typeScript / fileOutputs`.

Reply:
Upon reflection I realized there was an issue with the implementation. I updated the PR so that if you call (copyFile / allOutputPaths).value, it will invoke copyFile. Previously this was only the case if copyFile returned a File, Path or Seq of one or the other. You can see this by inspecting the graph:

sbt:nio-clean> inspect copyFile / allOutputPaths
[info] Task: scala.collection.Seq[java.nio.file.Path]
[info] Description:
[info]  All of the file output for a task with no filters applied.
[info] Provided by:
[info]  ProjectRef(uri("file:/Users/ethanatkins/work/scratch/nio-clean/"), "nio-clean") / copyFile / allOutputPaths
[info] Defined at:
[info]  (sbt.nio.Settings.allOutputPathsImpl) Settings.scala:286
[info] Dependencies:
[info]  copyFile / dynamicFileOutputs
[info]  ./ ./ ./ fileTreeView
[info]  copyFile / fileOutputs
[info] Reverse dependencies:
[info]  copyFile / outputFileStamps
[info] Delegates:
[info]  copyFile / allOutputPaths
[info]  allOutputPaths
[info]  ThisBuild / copyFile / allOutputPaths
[info]  ThisBuild / allOutputPaths
[info]  Zero / copyFile / allOutputPaths
[info]  Global / allOutputPaths
[info] Related:
[info]  Test / package / allOutputPaths
[info]  Test / packageSrc / allOutputPaths
[info]  Compile / packageSrc / allOutputPaths
[info]  Runtime / dependencyDot / allOutputPaths
[info]  Test / products / allOutputPaths
[info]  Test / productDirectories / allOutputPaths
[info]  Runtime / ivyReport / allOutputPaths
[info]  Provided / dependencyGraphMl / allOutputPaths
[info]  Test / ivyReport / allOutputPaths
[info]  Optional / ivyReport / allOutputPaths
[info] ...
sbt:nio-clean> inspect copyFile / dynamicFileOutputs
[info] Task: scala.collection.Seq[java.nio.file.Path]
[info] Description:
[info]  The outputs of a task
[info] Provided by:
[info]  ProjectRef(uri("file:/Users/ethanatkins/work/scratch/nio-clean/"), "nio-clean") / copyFile / dynamicFileOutputs
[info] Defined at:
[info]  /Users/ethanatkins/work/scratch/nio-clean/build.sbt:7
[info] Dependencies:
[info]  copyFile
[info] Reverse dependencies:
[info]  copyFile / allOutputPaths
[info] Delegates:
[info]  copyFile / dynamicFileOutputs
[info]  dynamicFileOutputs
[info]  ThisBuild / copyFile / dynamicFileOutputs
[info]  ThisBuild / dynamicFileOutputs
[info]  Zero / copyFile / dynamicFileOutputs
[info]  Global / dynamicFileOutputs
[info] Related:
[info]  Provided / ivyReport / dynamicFileOutputs
[info]  Compile / productDirectories / dynamicFileOutputs
[info]  Test / managedResources / dynamicFileOutputs
[info]  Test / products / dynamicFileOutputs
[info]  Test / doc / dynamicFileOutputs
[info]  Compile / package / dynamicFileOutputs
[info]  Compile / doc / dynamicFileOutputs
[info]  Compile / packageBin / dynamicFileOutputs
[info]  Compile / products / dynamicFileOutputs
[info]  Test / resources / dynamicFileOutputs
[info] ...
sbt:nio-clean> 
```
In particular note that copyFile / dynamicFileOutputs has a dependency on copyFile.